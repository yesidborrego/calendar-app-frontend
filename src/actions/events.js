import Swal from "sweetalert2";

import { convertDateEvents } from "../helpers/convertDateEvents";
import { fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";

export const eventSetActive = (event) => (
  {
    type: types.eventSetActive,
    payload: event
  }
);

export const eventStartAddNew = (event) => {
  return async (dispatch, getState) => {
    const { uid, name } = getState().auth;

    try {
      const res = await fetchWithToken('events', event, 'POST');
      const body = await res.json();

      event.id = body.evento.id;
      event.user = {
        _id: uid,
        name
      };

      dispatch(eventAddNew(event));

    } catch (error) {
      console.error(error);
    }
  }
}

const eventAddNew = (event) => (
  {
    type: types.eventAddNew,
    payload: event
  }
);

export const eventStartUpdate = (event) => {
  return async (dispatch) => {
    try {
      const res = await fetchWithToken(`events/${event.id}`, event, 'PUT');
      const body = await res.json();

      if(body.ok) {
        dispatch(eventUpdated(event));
      } else {
        Swal.fire('Error', body.msg, 'error');
      }

    } catch (error) {
      console.error(error);
    }
  }
}

const eventUpdated = (event) => (
  {
    type: types.eventUpdated,
    payload: event
  }
);

export const eventStartDelete = () => {
  return async (dispatch, getState) => {
    try {
      const { id } = getState().calendar.activeEvent;
      const res = await fetchWithToken(`events/${id}`, {}, 'DELETE');
      const body = await res.json();
      if(body.ok) {
        dispatch(eventDeleted());
      } else {
        Swal.fire('Error', body.msg, 'error');
      }

    } catch (error) {
      console.error(error);
    }
  }
}

const eventDeleted = () => (
  {
    type: types.eventDeleted,
  }
);

export const eventClearActive = () => (
  {
    type: types.eventClearActive,
  }
);

export const eventStartLoad = () => {
  return async (dispatch) => {
    try {
      const res = await fetchWithToken('events');
      const body = await res.json();

      const events = convertDateEvents(body.events);
      dispatch(eventLoaded(events));

    } catch (error) {
      console.log(error);
    }
  }
}

const eventLoaded = (events) => (
  {
    type: types.eventStartLoad,
    payload: events
  }
)

export const eventClearState = () => (
  {
    type: types.eventClearState
  }
)