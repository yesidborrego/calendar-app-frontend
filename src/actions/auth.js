import Swal from "sweetalert2";

import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch"
import { types } from "../types/types";
import { eventClearState } from "./events";

export const startLogin = (email, password) => {
  return async ( dispatch ) => {
    const res = await fetchWithoutToken('auth', {email, password}, 'POST');
    const body = await res.json();

    if( body.ok ) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime());

      dispatch(login({ uid: body.uid, name: body.name }));
    } else {
      Swal.fire('Error', body.msg, 'error');
    }
  }
}

export const startRegister = (email, password, name) => {
  return async (dispatch) => {
    const res = await fetchWithoutToken('auth/new', {email, password, name}, 'POST');
    const body = await res.json();

    if( body.ok ) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime());

      dispatch(login({ uid: body.uid, name: body.name }));
    } else {
      Swal.fire('Error', body.msg, 'error');
    }
  }
}

const login = (user) => (
  {
    type: types.authLogin,
    payload: user
  }
);

export const startTokenRenew = () => {
  return async (dispatch) => {
    const res = await fetchWithToken('auth/renew');
    const body = await res.json();

    if( body.ok ) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-date', new Date().getTime());

      dispatch(login({ uid: body.uid, name: body.name }));
    } else {
      dispatch(authCheckingFinish());
    }
  }
}

const authCheckingFinish = () => ({
  type: types.authStartTokenRenew
})

export const startLogout = () => {
  return (dispatch) => {
    localStorage.clear();
    dispatch(eventClearState());
    dispatch(logout());
  }
}

const logout = () => (
  {
    type: types.authLogout
  }
);