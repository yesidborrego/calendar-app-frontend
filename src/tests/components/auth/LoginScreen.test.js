import { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Swal from 'sweetalert2';

import { startLogin, startRegister } from '../../../actions/auth';
import { LoginScreen } from '../../../components/auth/LoginScreen';

import '@testing-library/jest-dom';

jest.mock('../../../actions/auth', () => ({
  startLogin: jest.fn(),
  startRegister: jest.fn()
}))

jest.mock('sweetalert2', () => ({
  fire: jest.fn()
}))

const middleware = [thunk];
const mockStore = configureStore(middleware);
const initialState = {};
const store = mockStore(initialState);

store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={ store }>
    <LoginScreen />
  </Provider>
);

describe('Prueba del componente <LoginScreen />', () => {

  beforeEach( () => {
    jest.clearAllMocks();
  });

  test('Debe mostrarse correctamente el componente', () => {
    expect(wrapper).toMatchSnapshot();
  })

  test('Debe llamar el dispatch del login', () => {
    wrapper.find("input[name='lEmail']").simulate('change', {
      target: {
        name: 'lEmail',
        value: 'isabella@test.com'
      }
    });

    wrapper.find("input[name='lPassword']").simulate('change', {
      target: {
        name: 'lPassword',
        value: '12345678'
      }
    });

    wrapper.find('form').at(0).prop('onSubmit')({
      preventDefault(){}
    });

    expect(startLogin).toHaveBeenLastCalledWith('isabella@test.com', '12345678');

  })

  test('No hay regsitro si las contrase침as no son iguales', () => {
    wrapper.find("input[name='rPassword']").simulate('change', {
      target: {
        name: 'rPassword',
        value: '123456789'
      }
    });

    wrapper.find("input[name='rPassword_confirm']").simulate('change', {
      target: {
        name: 'rPassword_confirm',
        value: '12345678'
      }
    });

    wrapper.find('form').at(1).prop('onSubmit')({
      preventDefault(){}
    });

    expect(startRegister).toHaveBeenCalledTimes(0); // Opci칩m 1
    expect(startRegister).not.toHaveBeenCalled(); // Opci칩n 2
    expect(Swal.fire).toHaveBeenCalledWith('Error', 'Password and confirmation password must be the same', 'error');

  });

  test('Regsitrar usuario si las contrase침as son iguales', () => {
    wrapper.find("input[name='rPassword']").simulate('change', {
      target: {
        name: 'rPassword',
        value: '12345678'
      }
    });

    wrapper.find("input[name='rPassword_confirm']").simulate('change', {
      target: {
        name: 'rPassword_confirm',
        value: '12345678'
      }
    });

    wrapper.find('form').at(1).prop('onSubmit')({
      preventDefault(){}
    });

    expect(startRegister).toHaveBeenCalledWith('vanesa@test.com', '12345678', 'Vanesa');
    expect(Swal.fire).not.toHaveBeenCalled();
  })
})
