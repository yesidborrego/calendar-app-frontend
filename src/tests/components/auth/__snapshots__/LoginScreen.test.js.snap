// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba del componente <LoginScreen /> Debe mostrarse correctamente el componente 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Ingreso
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="lPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Login"
          />
        </div>
      </form>
    </div>
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Registro
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rName"
            onChange={[Function]}
            placeholder="Nombre"
            type="text"
            value="Vanesa"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rEmail"
            onChange={[Function]}
            placeholder="Correo"
            type="email"
            value="vanesa@test.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword"
            onChange={[Function]}
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="rPassword_confirm"
            onChange={[Function]}
            placeholder="Repita la contraseña"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Crear cuenta"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
